packages/main/src/plugin/kubernetes-client.ts

```diff
@@ -1049,64 +1015,159 @@ export class KubernetesClient {
+   * on the context are patched, and any new resources are created.
+   *
+   * @param context a context
+   * @param filePath file system path to a YAML Kubernetes spec
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   */
+  async applyResourcesFromFile(context: string, filePath: string, namespace?: string): Promise<KubernetesObject[]> {
+    const manifests = await this.loadManifestsFromFile(filePath);
+    return this.syncResources(context, manifests, 'apply', namespace);
+  }
+
+  /**
+   * Apply a given yaml file to a context, i.e. 'kubectl apply -f'. Resources that exist
+   * on the context are patched, and any new resources are created.
+   *
+   * @param context a context
+   * @param filePath file system path to a YAML Kubernetes spec
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   */
+  async applyResources(
+    context: string,
+    manifests: KubernetesObject[],
+    namespace?: string,
+  ): Promise<KubernetesObject[]> {
+    return this.syncResources(context, manifests, 'apply', namespace);
+  }
+
+  /**
+   * Applies or creates a set of resources to a context, via creation or patching.
+   *
+   * @param context a context
+   * @param manifests a set of Kubernetes spec manifests
+   * @param action 'create' (only create new resources, do not overwrite existing) or 'apply' (create or patch as necessary)
+   * @param namespace the namespace to use for any resources that don't include one
+   * @return an array of resources created
+   *
+   * Heavily influenced by the API example:
+   * https://github.com/kubernetes-client/javascript/blob/0fbfd8fc2dcc7f4ec3e6fcd64a5c55169b6ef0b8/examples/typescript/apply/apply-example.ts
+   */
+  async syncResources(
+    context: string,
+    manifests: KubernetesObject[],
+    action: 'create' | 'apply',
+    namespace?: string,
+  ): Promise<KubernetesObject[]> {
+    const telemetryOptions: Record<string, unknown> = {
+      manifestsSize: manifests?.length,
+      action: action,
+    };
+    if (namespace) {
+      telemetryOptions.namespace = namespace;
+    }
+
     try {
       const ctx = new KubeConfig();
       ctx.loadFromFile(this.kubeconfigPath);
       ctx.currentContext = context;
-      for (const manifest of manifests) {
-        // https://github.com/kubernetes-client/javascript/issues/487
-        if (manifest?.metadata?.creationTimestamp) {
-          manifest.metadata.creationTimestamp = new Date(manifest.metadata.creationTimestamp);
+
+      const validSpecs = manifests.filter(s => s?.kind) as KubernetesObjectWithKind[];
+
+      const client = ctx.makeApiClient(KubernetesObjectApi);
+      const created: KubernetesObject[] = [];
+      for (const spec of validSpecs) {
+        // this is to convince TypeScript that metadata exists even though we already filtered specs
+        // without metadata out
```

Comment:
need to remove this comment as it's not filtered out

Reply:
Fixed.